app/models/concerns/unparsed_run.rb

```diff
@@ -13,154 +18,286 @@ def parse_into_db
+          Parser::LiveSplitCore::Run.parse_file_handle(f.fileno, '', false).with do |parse_result|
+            raise UnparsableRun unless parse_result.parsed_successfully
 
-        timer_used = Run.program_from_attribute(:to_s, parse_result[:program]).to_sym
+            run_data = process_run_wrapper(parse_result, run_data)
+            run = parse_result.unwrap
+          end
 
-        default_timing = Run::REAL
-        if parse_result[:realtime_duration_ms].zero? && parse_result[:gametime_duration_ms].positive?
-          default_timing = Run::GAME
+          run_data[:size] = File.size(f)
         end
 
-        populate_category(parse_result[:game], parse_result[:category]) if game.nil? || category.nil?
+        # Process and write run parts individually, and in batches to prevent loading N objects in memory at once
+        run_data = process_run_metadata(run, run_data)
+        process_run_history(run)
+        process_segments(run)
+        # Processing of segment histories is called from process_segments
 
-        splits = parse_result[:splits]
+        add_default_proc!(run_data)
 
-        write_run_histories(parse_result[:history]) if parse_result[:history].present?
-
-        write_segments(splits)
-        write_segment_histories(splits)
+        populate_category(run_data[:game], run_data[:category]) if game.nil? || category.nil?
 
         update(
           parsed_at:               Time.now.utc,
-          program:                 timer_used.to_s,
-          attempts:                parse_result[:attempts],
-          srdc_id:                 srdc_id || parse_result[:metadata][:srdc_id],
+          program:                 run_data[:program].to_s,
+          attempts:                run_data[:attempts],
+          srdc_id:                 srdc_id || run_data[:metadata][:srdc_id],
 
-          realtime_duration_ms:    zero_to_nil(parse_result[:realtime_duration_ms]),
-          realtime_sum_of_best_ms: zero_to_nil(parse_result[:realtime_sum_of_best_ms]),
+          realtime_duration_ms:    zero_to_nil(run_data[:realtime_duration_ms]),
+          realtime_sum_of_best_ms: zero_to_nil(run_data[:realtime_sum_of_best_ms]),
 
-          gametime_duration_ms:    zero_to_nil(parse_result[:gametime_duration_ms]),
-          gametime_sum_of_best_ms: zero_to_nil(parse_result[:gametime_sum_of_best_ms]),
+          gametime_duration_ms:    zero_to_nil(run_data[:gametime_duration_ms]),
+          gametime_sum_of_best_ms: zero_to_nil(run_data[:gametime_sum_of_best_ms]),
 
-          total_playtime_ms:       parse_result[:total_playtime_ms],
-          default_timing:          default_timing,
-          filesize_bytes:          size
+          total_playtime_ms:       run_data[:total_playtime_ms],
+          default_timing:          run_data[:default_timing],
+          filesize_bytes:          run_data[:size]
         )
 
         HighlightSuggestion.from_run(self)
       end
+    rescue UnparsableRun
+      destroy
```

Comment:
This is probably not the behavior we want atm, but wasn't sure exactly how we want to handle this right now since we could receive a request to parse before the hour is up in the presigned request.